enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/catchup/storecopy/LocalDatabase.java

```diff
@@ -193,4 +213,32 @@ public TransactionCommitProcess getCommitProcess()
     {
         return localCommit;
     }
+
+    private synchronized void stopWithRequirement( AvailabilityRequirement requirement ) throws Throwable
+    {
+        log.info( "Stopping, reason: " + requirement.description() );
+        databaseHealth = null;
+        localCommit = null;
+        dataSourceManager.stop();
+
+        raiseAvailabilityGuard( requirement );
```

Comment:
Yeah, it is definitely better to raise the guard before stopping the datasource. Fixed now.

I think if transaction gets in-between guard and `NeoStoreDataSource#stop()` it'll be asked to terminate during `NSDS#stop()`. Locks are termination aware so transactions should not just block somewhere.